### PR TITLE
Optimize IndexingContentHandlers perf

### PR DESCRIPTION
**Pull request by hyzx86**
try to fix https://github.com/OrchardCMS/OrchardCore/issues/13071

1. Avoid retrieving content from the database multiple times.

image

## Summary by CodeRabbit

- **New Features**
	- Enhanced content management capabilities with improved handling and indexing logic across different contexts such as publishing, updating, and creating content items.
- **Refactor**
	- Streamlined content indexing processes for Lucene, Azure AI, and Elasticsearch to boost efficiency and clarity.
- **Tests**
	- Expanded test coverage to include new utilities and indexing functionalities.

### CODE REVIEW

**Comment by hyzx86** [comment]
I haven't changed the rebuildindex function yet, so let's see if I got it right here

**Comment by Skrypt** [comment]
I'm surprised it is broken. Will test it and analyze the PR.

**Comment by Piedone** [comment]
Do you want to get back to this, @Skrypt?

**Comment by Skrypt** [comment]
Index Latest option is working from my tests on main branch that I just did so nothing is broken. I will analyze the PR.

File: src/OrchardCore.Modules/OrchardCore.Search.Lucene/Handler/LuceneIndexingContentHandler.cs
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Search.Lucene/Handler/LuceneIndexingContentHandler.cs:110]: I think here it should be the opposite. Here IndexLatest means that we want to index draft contentItems. We should need to delete by contentItemVersionId only when IndexLatest is set to true.

File: src/OrchardCore.Modules/OrchardCore.Search.Lucene/Handler/LuceneIndexingContentHandler.cs
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Search.Lucene/Handler/LuceneIndexingContentHandler.cs:106]: storeLuceneDocument

Fix typo

File: src/OrchardCore.Modules/OrchardCore.Search.Lucene/Handler/LuceneIndexingContentHandler.cs
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Search.Lucene/Handler/LuceneIndexingContentHandler.cs:90]: I like the idea of using the Contexts to avoid to retrieve these later on. Though, I think that we are retrieving them back here in the Handler in the DB because historically we had issue as it could have been altered by another handler. Though, maybe not an issue anymore.

File: src/OrchardCore.Modules/OrchardCore.Search.Lucene/Services/LuceneIndexManager.cs
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Search.Lucene/Services/LuceneIndexManager.cs:98]: Ok

**Review by Skrypt — DISMISSED** [review]

**Review by Skrypt — APPROVED** [review]
I think we need unit tests to make sure all scenarios work with the DefaultContentManager. Because, while testing this I needed to also fix the use case when we unpublish a content item which was not removing the content item in the index. I'm approving, but we need unit tests.

**Comment by Skrypt** [comment]
@MikeAlhayek It may need to be also updated for AzureAI and Elasticsearch.

**Comment by Skrypt** [comment]
I would love to have a perf comparison too. We need to have perf unit tests with Lucene. Populate 10 000 content items and index them. Something that should not be ran by the CI every time though. At least, we should have an easy way to populate 10 000 items to test this locally.

**Comment by Skrypt** [comment]
Don't merge yet. I will do the other ContentHandler in this PR while I'm at it.

File: src/OrchardCore.Modules/OrchardCore.Search.Lucene/Handler/LuceneIndexingContentHandler.cs
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Search.Lucene/Handler/LuceneIndexingContentHandler.cs:94]: so many changes here. It's hard to follow. can you more this local function outside to reduce the noise?

**Review by MikeAlhayek — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.Search.Lucene/Services/LuceneIndexManager.cs
Comment (Skrypt) [inline src/OrchardCore.Modules/OrchardCore.Search.Lucene/Services/LuceneIndexManager.cs:98]: This is unnecessary as we always only store one version of a content item (contentItemid) in the indexes (Lucene, Elasticsearch). And also the contentItemId is the only field that we know that will never be null.

Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.Search.Lucene/Services/LuceneIndexManager.cs]: removed

**Comment by Skrypt** [comment]
We need unit test at least for Lucene. Then it will be a copy for the other 2.

**Comment by Skrypt** [comment]
@hyzx86 Will you do the Unit Tests or should we create a ticket so that someone works on it?

**Comment by Piedone** [comment]
OK, let's do that, unless @hyzx86 you'd like to implement the tests.

**Comment by hyzx86** [comment]
@Piedone  how about this https://github.com/OrchardCMS/OrchardCore/pull/14063/files#diff-fbf089138615b7f7aa6d42d2792ab80878d4a50d3d17e0d2b9eb7e19439bc21b?

 We may need to make some adjustments to the verification logic of the data, and the verification method should be similar, right?

**Comment by Piedone** [comment]
It should perhaps be similar, yes, though for this, most possibly you can have a lower-level (more like unit, not integration) test.

File: src/OrchardCore.Modules/OrchardCore.Search.Lucene/Handler/LuceneIndexingContentHandler.cs
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Search.Lucene/Handler/LuceneIndexingContentHandler.cs:149]: StoreDocumentAsync

File: src/OrchardCore.Modules/OrchardCore.Search.Lucene/Handler/LuceneIndexingContentHandler.cs
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Search.Lucene/Handler/LuceneIndexingContentHandler.cs:152]: Create a local `new string[] { contentItem.ContentType }` and reuse the same instance in both method calls

File: src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Handlers/PublishContentContext.cs
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Handlers/PublishContentContext.cs:7]: Replace all `isPublishing == true` by `isPublishing`, in all the files of this PR

File: src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Handlers/PublishContentContext.cs
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Handlers/PublishContentContext.cs:5]: why nullable `bool?` ?

Comment (Skrypt) [reply src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Handlers/PublishContentContext.cs:5]: I will take a look at these comments @sebastienros 
I just wanted to have an optional param there so that we don't need to pass it every time.

Comment (sebastienros) [reply src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Handlers/PublishContentContext.cs:5]: The new `IsPublishing` worries me, the DefaultContentManager mentions that something can be null and handlers should take that into account, not create new dynamic fields like this. It's like saying we have a thing (Publish) but it's not really for that (Publishing)

Comment (Skrypt) [reply src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Handlers/PublishContentContext.cs:5]: And by that I mean on the DefaultContentManager.

Comment (Skrypt) [reply src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Handlers/PublishContentContext.cs:5]: Though, I did not change the behavior of what it was doing before. So, I'm affecting the PublishingItem only when we are actually publishing. I'm just moving it to a param instead of leaving it as a public setter. Because, that's what we don't want, to let people affect this PublishingItem other than on instanciation.

Comment (Skrypt) [reply src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Handlers/PublishContentContext.cs:5]: That or we should have an Unpublishing event

File: src/OrchardCore.Modules/OrchardCore.Search.Lucene/Handler/LuceneIndexingContentHandler.cs
Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.Search.Lucene/Handler/LuceneIndexingContentHandler.cs:152]: The other one is `new string[] { contentItem.ContentItemId }`

**Comment by hyzx86** [comment]
> @hyzx86 Will you do the Unit Tests? I created feature enhancement ticket here : #15601

ok, let me try.

**Comment by hyzx86** [comment]
Hi guys, I added a unit test to this code. My main purpose is to verify that the Lucene index only indexes the latest version

File: test/OrchardCore.Tests/Apis/ContentManagement/DeploymentPlans/ContentStepLuceneQueryTests.cs
Comment (hyzx86) [inline test/OrchardCore.Tests/Apis/ContentManagement/DeploymentPlans/ContentStepLuceneQueryTests.cs:118]: Using the `LuceneQuerySource` directly, avoid  to call `ContentManagement.GetAsync`

**Comment by Skrypt** [comment]
@hyzx86 Not forgetting, will try to find time to review this week.

File: test/OrchardCore.Tests/Apis/ContentManagement/DeploymentPlans/ContentStepLuceneQueryTests.cs
Comment (Skrypt) [inline test/OrchardCore.Tests/Apis/ContentManagement/DeploymentPlans/ContentStepLuceneQueryTests.cs:85]: Do we not want to test the DefaultContentManager instead then?

Comment (hyzx86) [reply test/OrchardCore.Tests/Apis/ContentManagement/DeploymentPlans/ContentStepLuceneQueryTests.cs:85]: Because the `DefaultContentManager` api always returns only the latest version

Comment (Skrypt) [reply test/OrchardCore.Tests/Apis/ContentManagement/DeploymentPlans/ContentStepLuceneQueryTests.cs:85]: Yes but we want to test the IndexingHandlers that are triggered by events from the DefaultContentManager. These indexing handlers should take care of indexing proper content item version.

Comment (hyzx86) [reply test/OrchardCore.Tests/Apis/ContentManagement/DeploymentPlans/ContentStepLuceneQueryTests.cs:85]: I'm confused, you mean we need to use `IContentManager` in this unit test to recheck whether the content items have been updated to the latest, right? Or other tests are needed

Comment (Skrypt) [reply test/OrchardCore.Tests/Apis/ContentManagement/DeploymentPlans/ContentStepLuceneQueryTests.cs:85]: Yes, I think what we want to test is what happens when a content items is created/drafted/updated with a user action with the IContentManager. Else, we are in the context of executing a background task. We don't have any unit tests for this either so what you did is not lost.

**Comment by Piedone** [comment]
Please check out https://github.com/OrchardCMS/OrchardCore/pull/14063 after merging this.

**Comment by Skrypt** [comment]
@coderabbitai review

**Comment by Piedone** [comment]
Does this still need triage, @Skrypt? It seems it was actually triaged, with an all-start reviewer team being there.

**Comment by Skrypt** [comment]
I believe it still needs proper unit tests. If we want to merge then we need to at least create a task/ticket/issue about Lucene Unit Tests.

**Comment by Piedone** [comment]
What do we need the triage for, exactly? Because then we don't merge the PR until it has unit tests.

**Comment by Skrypt** [comment]
Ok, so no merge untill we have proper unit tests just like we said before.

**Comment by hishamco** [comment]
@hyzx86 can you write a unit test for this to finalize this PR

Let me know if you need help

**Comment by hyzx86** [comment]
> @hyzx86 can you write a unit test for this to finalize this PR
> 
> Let me know if you need help

Yes，I'm waiting for the PR https://github.com/OrchardCMS/OrchardCore/pull/14063  merge first,

Over there I added a waiting task help class 

https://github.com/OrchardCMS/OrchardCore/pull/14063/files#diff-26c65b35ade31e00a358b01f3f4158f4824220dcf72aebce1b76ce84f8e47f5e

**Comment by hyzx86** [comment]
@hishamco, @Skrypt I need help ! 😅

Do we currently have a way to check that all indexing tasks in the system have been completed?
I noticed that there is a table called `IndexingTask`.  
Can we check this table to make sure that all tasks have been completed?

https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Indexing/IndexingTaskManager.cs

But I noticed that will always be some data in this table, I don't know if it is because the unit tests will not wait for 
`IndexingTaskManager.FlushAsync ` completes the reason, I will go to test

image

**Comment by hishamco** [comment]
As I looked at the code the `FlushAsync()` should handle this

> I don't know if it is because the unit tests will not wait for
IndexingTaskManager.FlushAsync  completes the reason, 

Which unit test do you refer to?

File: test/OrchardCore.Tests/Apis/ContentManagement/DeploymentPlans/ContentStepLuceneQueryTests.cs
Comment (hyzx86) [inline test/OrchardCore.Tests/Apis/ContentManagement/DeploymentPlans/ContentStepLuceneQueryTests.cs:127]: @hishamco , Here , 
Looks like I found another Bug, and here it will never work because of syntax errors

https://github.com/OrchardCMS/OrchardCore/blob/c967b5d49d7bc85ae3d58f32aa45a94cfee4f872/src/OrchardCore.Modules/OrchardCore.Indexing/IndexingTaskManager.cs#L133-L141

Comment (hyzx86) [reply test/OrchardCore.Tests/Apis/ContentManagement/DeploymentPlans/ContentStepLuceneQueryTests.cs:127]: ADO.net directly use string collections as arguments in the `in` expressions?

Comment (hyzx86) [reply test/OrchardCore.Tests/Apis/ContentManagement/DeploymentPlans/ContentStepLuceneQueryTests.cs:127]: Apparently not, the logic here is only to prevent repeated insertions, and there is no code to clean up the index task after the task is completed 😢

File: test/OrchardCore.Tests/Apis/Lucene/LuceneQueryTests.cs
Comment (hyzx86) [inline test/OrchardCore.Tests/Apis/Lucene/LuceneQueryTests.cs:205]: @Skrypt 
Our content item operation is a bit complicated~~

Comment (hyzx86) [reply test/OrchardCore.Tests/Apis/Lucene/LuceneQueryTests.cs:205]: Hi @Piedone ，need your help , I don't know what to do..
https://github.com/OrchardCMS/OrchardCore/issues/15601#issuecomment-2090289841

Comment (Piedone) [reply test/OrchardCore.Tests/Apis/Lucene/LuceneQueryTests.cs:205]: What's your question, specifically?

Comment (hyzx86) [reply test/OrchardCore.Tests/Apis/Lucene/LuceneQueryTests.cs:205]: Just as @Skrypt  mentioned in this issue at https://github.com/OrchardCMS/OrchardCore/issues/15601 ,
I understand that we should verify the impact on indexing functionality by separately testing Create, Update, Draft, and Publish actions with ContentManager. 

However, based on the current design of `DefaultContentManager` which requires calling a sequence of methods to save data to the database and trigger Lucene indexing, it seems difficult if not impossible to isolate tests for just the Create and Update actions.

All that aside, am I currently unit testing the right way?

File: test/OrchardCore.Tests/Utilities/TimeoutTaskRunner.cs
Comment (hyzx86) [reply test/OrchardCore.Tests/Utilities/TimeoutTaskRunner.cs:12]: Please offer some suggestions

File: test/OrchardCore.Tests/Apis/Lucene/LuceneQueryTests.cs
Comment (Piedone) [reply test/OrchardCore.Tests/Apis/Lucene/LuceneQueryTests.cs:205]: You don't need to test `DefaultContentManager` or have a full indexing test. Rather, unit tests on `LuceneIndexingContentHandler` with mocks is sufficient. I.e, the tests need to check that `LuceneIndexingContentHandler` behaves correctly based on what it calls on the mocks.

If you search for "mock.setup" and "mock.verify" in the codebase, you'll see how this works. A lot of tests do something similar.

**Pull request closed without merging**
